source/thead_rvv/fp32/prelu.c

```diff
@@ -53,21 +54,28 @@ int shl_rvv_prelu_fp32(struct csinn_tensor *input, struct csinn_tensor *alpha,
             output->layout = CSINN_LAYOUT_NC1HWC0;
         }
     } else if (input->layout == CSINN_LAYOUT_NCHW) {
-        for (int n = 0; n < input->dim[0]; ++n) {
-            for (int c = 0; c < input->dim[1]; ++c) {
-                float a = alpha_data[c];
-                int inner_size = input->dim[2] * input->dim[3];
-                while (inner_size > 0) {
-                    int vl = vsetvl_e32m2(inner_size);
-                    vfloat32m2_t _input = vle32_v_f32m2(input_data, vl);
-                    vbool16_t _mask = vmflt_vf_f32m2_b16(_input, 0.0f, vl);
-                    vfloat32m2_t _res = vfmul_vf_f32m2_m(_mask, _input, _input, a, vl);
-                    vse32_v_f32m2(output_data, _res, vl);
-                    input_data += vl;
-                    output_data += vl;
-                    inner_size -= vl;
-                }
-            }
+        // broadcast alpha
+        int new_alpha_size = csinn_tensor_size(input);
+        float *alpha_data_b = shl_mem_alloc(new_alpha_size * sizeof(float));
+        struct csinn_tensor *alpha_ = csinn_alloc_tensor(NULL);
+        csinn_tensor_copy(alpha_, input);
+        alpha_->data = alpha_data_b;
+        shl_ref_broadcast_to_shape_f32(alpha, alpha_, alpha_->dim, alpha_->dim_count);
+        alpha_data = (float *)alpha_->data;
+
+        // calculation
+        int ele_size = input->dim[0] * input->dim[1] * input->dim[2] * input->dim[3];
+        while (ele_size > 0) {
+            int vl = vsetvl_e32m2(ele_size);
+            vfloat32m2_t _input = vle32_v_f32m2(input_data, vl);
+            vfloat32m2_t _a = vle32_v_f32m2(alpha_data, vl);
+            vbool16_t _mask = vmflt_vf_f32m2_b16(_input, 0.0f, vl);
+            vfloat32m2_t _res = vfmul_vv_f32m2_m(_mask, _input, _input, _a, vl);
+            vse32_v_f32m2(output_data, _res, vl);
+            input_data += vl;
+            alpha_data += vl;
+            output_data += vl;
+            ele_size -= vl;
```

Comment:
In models with PRelu op the second input `alpha` of this op has shape `[1]` (scalar) or `[1,C,1,1]` (per channel).
As I can see, SHL efficiently supports this op for these cases (via LeakyRelu and PRelu).
So when you call `shl_ref_broadcast_to_shape_f32` each time for `[1,C,1,1]`, it may lead to great performance degradations.
If we want to support broadcasting (not usual case but possible), we can just check if `alpha->dim[1] != csinn_tensor_size(alpha) == true` - need to call your new code. Otherwise need to call SHL-implemented code.
Thus, we will have:
1. Efficient and optimized implementation for frequent case in models - it's very important for fast inference
2. And support of any `alpha` shape for rare cases in models - it will be non efficient but supported.
What do you think?

Reply:
I agree that broadcast might cause significant overhead when executing. I think it is a really great way to add some checks in order to improve kernel performance! :)